<!DOCTYPE html>
<!-- saved from url=(0032)http://localhost:1313/resources/ -->
<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><script src="./Resources - ARGOS DB 2_files/livereload.js" data-no-instant="" defer=""></script>
    
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>
       Resources - ARGOS DB 2 
    </title>
    <meta name="HandheldFriendly" content="True">
    <meta name="MobileOptimized" content="320">
    <meta name="referrer" content="no-referrer">
    <meta name="description" content="">
    <meta property="og:site_name" content="ARGOS DB 2">
    <meta property="og:locale" content="en_US">
    <meta property="og:type" content="article">
    <meta property="og:url" content="//localhost:1313/resources/">
    <meta property="og:title" content="Resources">
    <meta property="og:image" content="//localhost:1313/">
    <meta property="og:description" content="">

    <meta name="twitter:card" content="summary_large_image"> 
    <meta name="twitter:title" content="Resources">
    <meta name="twitter:description" content="">
    <meta name="twitter:image" content="//localhost:1313/">

    <link rel="canonical" href="http://localhost:1313/resources/">

    <link rel="stylesheet" href="./Resources - ARGOS DB 2_files/bootstrap.min.css" integrity="sha512-MoRNloxbStBcD8z3M/2BmnT+rg4IsMxPkXaGh2zD6LGNNFE80W3onsAhRcMAMrSoyWL9xD7Ert0men7vR8LUZg==" crossorigin="anonymous">

    <link href="./Resources - ARGOS DB 2_files/css2" rel="stylesheet">
    <link rel="stylesheet" href="./Resources - ARGOS DB 2_files/custom.css">

     

    <link rel="shortcut icon" href="http://localhost:1313/images/favicon.png">

    
    <link href="http://localhost:1313/index.xml" rel="alternate" type="application/rss+xml" title="ARGOS DB 2">
    
  </head>

  <body>
     <div class="my-4 my-lg-5 header">
  <div class="container">
    <div class="row d-flex justify-content-center">
      <div class="col-auto">
        <ul class="nav nav-primary">
          
          <li class="nav-item">
            <a class="nav-link text-home" href="http://localhost:1313/">
               Home
            </a>
          </li>
          
          <li class="nav-item">
            <a class="nav-link text-about" href="http://localhost:1313/about/">
               About
            </a>
          </li>
          
          <li class="nav-item">
            <a class="nav-link text-resources" href="http://localhost:1313/resources/">
               Resources
            </a>
          </li>
           
        </ul>
      </div>
    </div>
  </div>
</div>


<div class="content">
  <div class="container">
    <div class="row justify-content-center">
      <div class="col-lg-12">
        <div class="markdown">
    <div class="col-lg-6 offset-lg-3 text-center">
<img src="./Resources - ARGOS DB 2_files/logo.about.png" class="img-fluid mx-auto d-block" width="75%" alt="">
</div>
<br>
<p>To cite BioCompute, please use:</p>
<ol>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/27974626/">Biocompute Objects-A Step towards Evaluation and Validation of Biomedical Scientific Computations</a><br></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/30596645/">Enabling precision medicine via standard communication of HTS provenance, analysis, and results</a></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/33784373/">Bioinformatics tools developed to support BioCompute Objects</a>
<br></li>
</ol>
<h3 id="table-of-contents">Table of contents</h3>
<p><a href="http://localhost:1313/resources/#BCO%20Portal">BCO Portal</a> <br>
<a href="http://localhost:1313/resources/#Platform%20Specific%20Tools">Platform Specific Tools</a> <br>
<a href="http://localhost:1313/resources/#BCO%20Community">BCO Community</a> <br>
<a href="http://localhost:1313/resources/#BCO%20Documentation">BCO Documentation</a> <br>
<a href="http://localhost:1313/resources/#Software%20Packages">Software Packages</a> <br>
<a href="http://localhost:1313/resources/#BCO%20Registry">BCO Registry</a> <br>
<a href="http://localhost:1313/resources/#BCO%20Aggregator">BCO Aggregator</a> <br>
<a href="http://localhost:1313/resources/#BCO%20Cheat%20Sheet">BCO Cheat Sheet</a> <br>
<a href="http://localhost:1313/resources/#References">References</a><br>
<a href="http://localhost:1313/events.md">Previous Presentations</a></p>
<br>
<p>Publicly available resources developed by the BioCompute community include:</p>
<p><a name="BCO Portal"></a></p>
<h3 id="bco-portal">BCO Portal</h3>
<p>The <a href="https://portal.aws.biochemistry.gwu.edu/">BCO Portal</a> is a form based creation tool for generating BioCompute Objects. The BCO Portal contains built-in linting and conformance against the current BioCompute specification. The BCO Portal includes a database of created BCOs. Unless specified in the embargo field, BCOs in the database are viewable to the general public.</p>
<p><a name="Platform Specific Tools"></a></p>
<h3 id="platform-specific-tools">Platform Specific Tools</h3>
<h4 id="hive-bco-app">HIVE BCO App</h4>
<p>The High-throughput Integrated Virtual Environment (<a href="https://hive.biochemistry.gwu.edu/dna.cgi?cmd=main">HIVE</a>) for genome analysis and the widely used, open source Galaxy project both have platform specific tools for generating BioCompute Objects from workflows. Both of these can be found on our <a href="http://localhost:1313/#tools">home page</a>.</p>
<h4 id="cancer-genomics-cloud-bco-app">Cancer Genomics Cloud BCO App</h4>
<p>The <a href="https://www.cancergenomicscloud.org/">Cancer Genomics Cloud</a> (CGC) has a powerful tool built into the CGC platform to capture and export a workflow as a BioCompute Object.</p>
<h4 id="galaxy-bco-extension">Galaxy BCO Extension</h4>
<p>The <a href="https://galaxy.aws.biochemistry.gwu.edu/">Galaxy BCO API Extension</a> enables Galaxy users to the export of Galaxy “workflow invocations” (i.e. realizations of a computational pipeline) in BCO format.</p>
<p><a name="BCO Community"></a></p>
<h3 id="bco-community">BCO Community</h3>
<ul>
<li>The <a href="https://gitter.im/biocompute-objects/">BCO Gitter</a> has spaces for general discussion, technical development, BCO implementation, and networking.</li>
</ul>
<p><a name="BCO Documentation"></a></p>
<h3 id="bco-documentation">BCO Documentation</h3>
<ul>
<li>The <a href="http://localhost:1313/BCO_SOP">BCO SOP</a> provides guidance on BCO creation, versioning, and certification.</li>
<li>The <a href="http://localhost:1313/BCO_UserManual">BCO User Manual</a> provides guidance on BCO evaluation.</li>
<li>The <a href="https://biocompute-objects.github.io/bco-ro-crate/">BCO RO Tutorial</a> shows how a BCO can be packaged/archived/transmitted/described as an RO-Crate Research Object, using a running <a href="http://localhost:1313/resources/github.com/biocompute-objects/bco-ro-example-chipseq">example of a Nextflow workflow</a>. <strong>Work in progress</strong></li>
</ul>
<p><a name="Software Packages"></a></p>
<h3 id="software-packages">Software Packages</h3>
<ul>
<li>The R package <a href="https://cran.r-project.org/package=biocompute">biocompute</a> can create, validate, and export BioCompute Objects. Users can encode information in data frames, and compose BioCompute Objects from the domains defined by the standard. A checksum validator and a JSON schema validator are provided. Also supports exporting BioCompute Objects as JSON, PDF, HTML, or Word documents, and exporting to cloud-based platforms.</li>
</ul>
<p><a name="BCO Registry"></a></p>
<h3 id="bco-registry">BCO Registry</h3>
<ul>
<li>The BioCompute consortium maintains a database of <a href="https://biocomputeobject.org/registry.html">registered authorities</a>. Registered authorities are able to assign their own IDs in the <code>object_id</code> field, such as gwu000001.</li>
</ul>
<p><a name="BCO Aggregator"></a></p>
<h3 id="bco-aggregator-under-construction">BCO Aggregator (under construction)</h3>
<ul>
<li>The BioCompute consortium maintains a database of submitted BCOs. The database acts as a mirror that generates checksums for validation.</li>
</ul>
<p><a name="BCO Cheat Sheet"></a></p>
<h3 id="bco-cheat-sheet">BCO “Cheat Sheet”</h3>
<ul>
<li>A <a href="http://localhost:1313/BCOCheatSheet.pdf">quick reference guide</a> (PDF) for getting up to speed with reading a BCO.</li>
</ul>
<p><a name="References"></a></p>
<h3 id="references">References</h3>
<ol>
<li>
<p><a href="https://opensource.ieee.org/2791-object/ieee-2791-schema/">2791-2020 IEEE Standard</a><br>
The official IEEE standard, known as IEEE 2791-2020.</p>
</li>
<li>
<p><a href="https://standards.ieee.org/standard/2791-2020.html">Repository</a><br>
The open source repository referenced by the standard. Three formats of the publication are available.</p>
</li>
<li>
<p><a href="https://semver.org/spec/v2.0.0.html">Semantic Versioning</a><br>
BioCompute versioning is based on semantic versioning, but does not use the “major” value in “major.minor.patch” (e.g. <code>2.4.11</code>).</p>
<ul>
<li>Use “minor” increments for additions (e.g. the addition of a “<code>reviewedBy</code>” block).</li>
<li>Use “patch” for minor fixes (e.g. editorial spelling corrections).</li>
<li>There is no “major” increment in BCO versioning. A BCO is a record of a specific instance of an analysis, so major changes result in a new BCO.</li>
</ul>
</li>
<li>
<p><a href="https://jbiomedsem.biomedcentral.com/articles/10.1186/2041-1480-4-37">Provenance Ontology</a><br>
Contributor roles in a BCO are explicitly defined according to the Prov-O standard, and include roles for reviewing a BCO, creating the content in a BCO, and for authoring the BCO itself.</p>
</li>
<li>
<p><a href="https://json-schema.org/draft/2019-09/json-schema-core.html">JSON Schema</a><br>
BioCompute is based heavily on the JSON Schema media type. Definitions and format within the BCO Schema adhere to “draft-handrews-json-schema-02.”</p>
</li>
</ol>


</div>
      </div>
    </div>
  </div>
</div>

<div class="my-4 footer">
  <div class="container">
    <div class="row justify-content-center">
      <div class="col-lg-12">
        <hr>
      </div>
    </div>

    <div class="row justify-content-center">
      <div class="col-sm-12 col-lg-6">
        
      </div>
      <div class="col-sm-12 col-lg-6">
        <div class="mx-0 mx-md-4 site-social">
          <ul>
            
            <li><a href="mailto:ssingleton@gwu.edu?cc=mazumder@gwu.edu?subjsect=Mail%20from%20Our%20Site">Contact</a></li>
            
             
            <li>
              <a href="https://github.com/FDA-ARGOS2" target="_blank">GitHub</a>
            </li>
                 
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
      
  

<script>mendeleyWebImporter = {
  downloadPdfs(t,e) { return this._call('downloadPdfs', [t,e]); },
  open() { return this._call('open', []); },
  setLoginToken(t) { return this._call('setLoginToken', [t]); },
  _call(methodName, methodArgs) {
    const id = Math.random();
    window.postMessage({ id, token: '0.2530653373944489', methodName, methodArgs }, 'http://localhost:1313');
    return new Promise(resolve => {
      const listener = window.addEventListener('message', event => {
        const data = event.data;
        if (typeof data !== 'object' || !('result' in data) || data.id !== id) return;
        window.removeEventListener('message', listener);
        resolve(data.result);
      });
    });
  }
};</script></body></html>